pkg/signer/memory.go

```diff
@@ -97,11 +114,33 @@ func NewTimestampingCertWithSelfSignedCA(pub crypto.PublicKey) ([]*x509.Certific
 		},
 		BasicConstraintsValid: true,
 	}
-	certBytes, err := x509.CreateCertificate(rand.Reader, cert, ca, pub.(*ecdsa.PublicKey), caPrivKey)
+
+	// Create the certificate
+	certBytes, err := x509.CreateCertificate(rand.Reader, cert, chain[0], pub, cryptoSigner)
 	if err != nil {
+		return nil, errors.Wrap(err, "creating tsa certificate")
+	}
+	tsaCert, err := x509.ParseCertificates(certBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify and return the certificate chain
+	root := x509.NewCertPool()
+	root.AddCert(chain[len(chain)-1])
+	intermediates := x509.NewCertPool()
+	for _, intermediate := range chain[:len(chain)-1] {
+		intermediates.AddCert(intermediate)
+	}
+	verifyOptions := x509.VerifyOptions{
+		Roots:         root,
+		Intermediates: intermediates,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
```

Comment:
can we constrain to timestamps here?

Reply:
Oops, done! I don't know why I missed it in https://cs.opensource.google/go/go/+/go1.16.6:src/crypto/x509/x509.go;l=570